src/ucp/proto/proto_select.c

```diff
@@ -689,34 +689,34 @@ void ucp_proto_select_short_disable(ucp_proto_select_short_t *proto_short)
     proto_short->rkey_index             = UCP_NULL_RESOURCE;
 }
 
-void
-ucp_proto_select_short_init(ucp_worker_h worker, ucp_proto_select_t *proto_select,
-                            ucp_worker_cfg_index_t ep_cfg_index,
-                            ucp_worker_cfg_index_t rkey_cfg_index,
-                            ucp_operation_id_t op_id, uint32_t op_attr_mask,
-                            unsigned proto_flags,
-                            ucp_proto_select_short_t *proto_short)
+void ucp_proto_select_short_init(ucp_worker_h worker,
+                                 ucp_proto_select_t *proto_select,
+                                 ucp_worker_cfg_index_t ep_cfg_index,
+                                 ucp_worker_cfg_index_t rkey_cfg_index,
+                                 ucp_operation_id_t op_id, unsigned proto_flags,
+                                 ucp_proto_select_short_t *proto_short)
 {
-    ucp_context_h context    = worker->context;
-    const ucp_proto_t *proto = NULL;
+    static const uint32_t op_attributes[] = {0, UCP_OP_ATTR_FLAG_FAST_CMPL,
+                                             UCP_OP_ATTR_FLAG_MULTI_SEND};
+    ucp_context_h context                 = worker->context;
+    const ucp_proto_t *proto              = NULL;
     const ucp_proto_threshold_elem_t *thresh;
     ucp_proto_select_param_t select_param;
     const ucp_proto_single_priv_t *spriv;
     ucp_memory_info_t mem_info;
     ssize_t max_short_signed;
-    uint32_t op_attr;
+    int i;
 
     ucp_memory_info_set_host(&mem_info);
 
     /*
-     * Find the minimal threshold among all protocols for all possible
-     * combinations of bits in 'op_attr_mask'. For example, we are allowed to
-     * use fast-path short protocol only if the message size fits short protocol
-     * in both regular mode and UCP_OP_ATTR_FLAG_FAST_CMPL mode.
+     * Find the minimal threshold for operation 'op_id' among all protocols
+     * with attribute from 'op_attributes'. Fast-path short protocol
+     * can be used only if the message size fits this minimal threshold.
      */
     ucs_log_indent(1);
-    ucs_for_each_submask(op_attr, op_attr_mask) {
-        ucp_proto_select_param_init(&select_param, op_id, op_attr, 0,
+    for (i = 0; i < ucs_static_array_size(op_attributes); ++i) {
+        ucp_proto_select_param_init(&select_param, op_id, op_attributes[i], 0,
```

Comment:
You check the threshold only when the flags are used individually.
Is there a use case when we don't want to check `UCP_OP_ATTR_FLAG_FAST_CMPL | UCP_OP_ATTR_FLAG_MULTI_SEND`?

And for me, it isn't still clear, why we previously (or originally) pass `UCP_OP_ATTR_FLAG_FAST_CMPL` only for put operation (in this PR https://github.com/openucx/ucx/pull/5850).

Maybe there was a reason for this. And it become clear, what combination of flags we should use during short protocol threshold calculation.

Reply:
I'm not sure if there is a use case. I think it's more related to meaning of each opt attribute, one prefers latency the other bandwidth. In that case I'm not sure why we need to check `UCP_OP_ATTR_FLAG_FAST_CMPL` if we check `UCP_OP_ATTR_FLAG_MULTI_SEND`. Seems to me that if you optimize bandwidth over latency short threshold will always be smaller, but it's just an assumption.

I'm not sure why we pass `UCP_OP_ATTR_FLAG_FAST_CMPL` only for put operation, but here are some basic facts:
- `UCP_OP_ATTR_FLAG_MULTI_SEND` was added after `UCP_OP_ATTR_FLAG_FAST_CMPL`
- The call to `ucp_proto_select_short_init` for AM/TAG was added after the call for PUT

So maybe it's just a bug, but lets wait for others to reply.

Reply:
`UCP_OP_ATTR_FLAG_MULTI_SEND` is for optimizing protocol selection in favor of bandwidth. However, I don't agree that `UCP_OP_ATTR_FLAG_FAST_CMPL` has the opposite meaning, and intendent for optimizing protocol selection in favor of latency. Otherwise, there is no need to have both flags if they have opposite effect. Thus, I don't see reasons, why we cannot use both flags in the same time for the protocol selection, and calculation of thresholds in particular.

Agree that `UCP_OP_ATTR_FLAG_MULTI_SEND` was added after `UCP_OP_ATTR_FLAG_FAST_CMPL`. But the next fact is not completely correct. The call to `ucp_proto_select_short_init` only for AM (https://github.com/openucx/ucx/pull/7644) was added after the call for TAG and PUT (https://github.com/openucx/ucx/pull/5850).

@yosefe do you remember, why did you pass `UCP_OP_ATTR_FLAG_FAST_CMPL` only in case of short put in this PR https://github.com/openucx/ucx/pull/5850?

```C
        /* Set threshold for short send */
        ucp_proto_select_short_init(worker, &ep_config->proto_select,
                                    ep_cfg_index, UCP_WORKER_CFG_INDEX_NULL,
                                    UCP_OP_ID_TAG_SEND, 0,
                                    UCP_PROTO_FLAG_AM_SHORT, &tag_short);
...
       /* Set threshold for short put */
        ucp_proto_select_short_init(worker, &rkey_config->proto_select,
                                    key->ep_cfg_index, rkey_cfg_index,
                                    UCP_OP_ID_PUT, UCP_OP_ATTR_FLAG_FAST_CMPL,
                                    UCP_PROTO_FLAG_PUT_SHORT,
                                    &rkey_config->put_short);
```

Reply:
FAST_CMPL usually implies blocking operation, which contradicts MULTI_SEND.

> @yosefe do you remember, why did you pass UCP_OP_ATTR_FLAG_FAST_CMPL only in case of short put in this PR https://github.com/openucx/ucx/pull/5850?

It was wrong, i didn't consider the fast-path threshold would be too high.
IMO the right thing is for fast-path threshold to be the minimum of all valid flag combinations.

Reply:
Yossi what exactly do you mean by "valid flag"?
Is it a combination you consider as valid or any flag combination that user can pass to UCX?

Reply:
A combination that "makes sense". IMO we can document (and validate with ENABLE_PARAM_CHECK) that UCP_OP_ATTR_FLAG_FAST_CMPL is not used along with UCP_OP_ATTR_FLAG_MULTI_SEND

Reply:
I think we need to document it and validate it at least with ENABLE_PARAM_CHECK so it will be well defined.
I think it also bothered Raul that we treat this combination as not Valid but we are not validating it or document it.

WDIT @rakhmets ?

Reply:
The changes are good for me in this case.